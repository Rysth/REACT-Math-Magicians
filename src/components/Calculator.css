:root {
  --CL-primary-orange: #f4903e;
  --CL-primary-light: #e0e1e1;
  --CL-primary-gray: #848692;
  --CL-neutral-white: #fff;
  --CL-neutral-black: #000;
}

.calculator-title {
  text-align: center;
  font-size: 1.8rem;
  color: var(--CL-secondary);
  margin: initial auto 0 auto;
}

.calculator {
  outline: 1px solid var(--CL-primary-light);
  max-width: 500px;
  margin: 3em auto;
}

.calculator-body {
  display: grid;
}

.calculator-result {
  padding: 0.5em;
  width: 100%;
  background-color: var(--CL-primary);
  color: white;
  outline: 1px solid var(--CL-primary-light);
  text-align: right;
  pointer-events: none;
  font-size: 2rem;
}

.calculator-result.active {
  background-color: var(--CL-neutral-black);
}

.calculator-group {
  display: inherit;
  grid-template-columns: repeat(4, 1fr);
}

.calculator-button {
  padding: 1em 0;
  cursor: pointer;
  outline: 1px solid var(--CL-primary-light);
  transition: 0.15s;
  font-size: 0.95rem;
}

.calculator-button.operation {
  background-color: var(--CL-secondary);
  color: var(--CL-neutral-white);
}

.calculator-button.expanded {
  grid-column: 1 / 3;
}

.calculator-button:hover {
  background-color: var(--CL-primary-orange);
  transform: scale(1.05);
}

.calculator-button:active {
  transform: scale(0.95);
}

@media (min-width: 768px) {
  .calculator-result {
    font-size: 2.25rem;
  }

  .calculator-button {
    font-size: 1.2rem;
  }
}
