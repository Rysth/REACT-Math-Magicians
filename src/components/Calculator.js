import PropTypes from 'prop-types';
import './Calculator.css';

const calculatorButtons = [
  { id: 0, classValue: '', content: 'AC' },
  { id: 1, classValue: '', content: '+/-' },
  { id: 2, classValue: '', content: '%' },
  { id: 3, classValue: 'operation', content: 'รท' },
  { id: 4, classValue: '', content: '7' },
  { id: 5, classValue: '', content: '8' },
  { id: 6, classValue: '', content: '9' },
  { id: 7, classValue: 'operation', content: 'x' },
  { id: 8, classValue: '', content: '4' },
  { id: 9, classValue: '', content: '5' },
  { id: 10, classValue: '', content: '6' },
  { id: 11, classValue: 'operation', content: '-' },
  { id: 12, classValue: '', content: '1' },
  { id: 13, classValue: '', content: '2' },
  { id: 14, classValue: '', content: '3' },
  { id: 15, classValue: 'operation', content: '+' },
  { id: 16, classValue: 'expanded', content: '0' },
  { id: 17, classValue: '', content: '.' },
  { id: 18, classValue: 'operation', content: '=' },
];

export default function Calculator() {
  return (
    <div className="calculator">
      <CalculatorResult />
      <div className="calculator-body">
        <CalculatorGroup array={calculatorButtons.slice(0, 4)} />
        <CalculatorGroup array={calculatorButtons.slice(4, 8)} />
        <CalculatorGroup array={calculatorButtons.slice(8, 12)} />
        <CalculatorGroup array={calculatorButtons.slice(12, 16)} />
        <CalculatorGroup array={calculatorButtons.slice(16, 19)} />
      </div>
    </div>
  );
}

function CalculatorResult({ result }) {
  return <input className="calculator-result" type="text" value={result} readOnly />;
}

CalculatorResult.defaultProps = {
  result: 0,
};

CalculatorResult.propTypes = {
  result: PropTypes.number,
};

function CalculatorGroup({ array }) {
  return (
    <div className="calculator-group">
      {array.map((element) => (
        <CalculatorButton
          key={element.id}
          classValue={element.classValue}
          content={element.content}
        />
      ))}
    </div>
  );
}

CalculatorGroup.defaultProps = {
  array: [],
};

CalculatorGroup.propTypes = {
  array: PropTypes.arrayOf(PropTypes.objectOf),
};

function CalculatorButton({ content, classValue }) {
  const classList = `calculator-button ${classValue}`;

  return (
    <button type="button" className={classList}>
      {content}
    </button>
  );
}

CalculatorButton.defaultProps = {
  content: '',
  classValue: '',
};

CalculatorButton.propTypes = {
  content: PropTypes.string,
  classValue: PropTypes.string,
};
