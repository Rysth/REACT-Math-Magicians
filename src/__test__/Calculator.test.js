import renderer from 'react-test-renderer';
import Calculator from '../components/Calculator';

describe('Calculator Component', () => {
  it('Should create a Calculator Component Snapshot', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
