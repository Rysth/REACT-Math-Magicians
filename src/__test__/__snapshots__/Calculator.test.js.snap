// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator Component Should create a Calculator Component Snapshot 1`] = `
Array [
  <div>
    <h3
      className="calculator-title"
    >
      Let's do some Math
    </h3>
  </div>,
  <div
    className="calculator"
  >
    <input
      className="calculator-result"
      readOnly={true}
      type="text"
      value={0}
    />
    <div
      className="calculator-body"
    >
      <div
        className="calculator-group"
      >
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          AC
        </button>
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          +/-
        </button>
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          %
        </button>
        <button
          className="calculator-button operation"
          onClick={[Function]}
          type="button"
        >
          รท
        </button>
      </div>
      <div
        className="calculator-group"
      >
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          7
        </button>
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          8
        </button>
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          9
        </button>
        <button
          className="calculator-button operation"
          onClick={[Function]}
          type="button"
        >
          x
        </button>
      </div>
      <div
        className="calculator-group"
      >
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          4
        </button>
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          5
        </button>
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          6
        </button>
        <button
          className="calculator-button operation"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
      <div
        className="calculator-group"
      >
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          1
        </button>
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          2
        </button>
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          3
        </button>
        <button
          className="calculator-button operation"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
      </div>
      <div
        className="calculator-group"
      >
        <button
          className="calculator-button expanded"
          onClick={[Function]}
          type="button"
        >
          0
        </button>
        <button
          className="calculator-button "
          onClick={[Function]}
          type="button"
        >
          .
        </button>
        <button
          className="calculator-button operation"
          onClick={[Function]}
          type="button"
        >
          =
        </button>
      </div>
    </div>
  </div>,
]
`;
